src/main/java/cms/gov/madie/measure/utils/QiCoreJsonUtil.java

```diff
@@ -0,0 +1,104 @@
+    ObjectMapper mapper = new ObjectMapper();
+    try {
+      mapper.readTree(json);
+      return true;
+    } catch (JsonProcessingException e) {
+      // do nothing
+    }
+    return false;
+  }
+
+  public static String getPatientId(String json) throws JsonProcessingException {
+    ObjectMapper mapper = new ObjectMapper();
+
+    JsonNode jsonNode = mapper.readTree(json);
+    JsonNode entry = jsonNode.get("entry");
+    JsonNode theNode = null;
+    Iterator<JsonNode> entyIter = entry.iterator();
+    String existingPatientId = null;
+    while (entyIter.hasNext()) {
+      theNode = entyIter.next();
+      var resourceNode = theNode.get("resource");
+      if (resourceNode != null) {
+        var resourceType = resourceNode.get("resourceType");
+        if (resourceType != null && "PATIENT".equalsIgnoreCase(resourceType.asText())) {
+          existingPatientId = resourceNode.get("id").textValue();
+        }
+      }
+    }
+    return existingPatientId;
+  }
+
+  public static String getPatientFullUrl(String json) throws JsonProcessingException {
+    ObjectMapper mapper = new ObjectMapper();
+
+    JsonNode jsonNode = mapper.readTree(json);
+    JsonNode entry = jsonNode.get("entry");
+    JsonNode theNode = null;
+    Iterator<JsonNode> entyIter = entry.iterator();
+    String fullUrl = null;
+    while (entyIter.hasNext()) {
+      theNode = entyIter.next();
+      var resourceNode = theNode.get("resource");
+      if (resourceNode != null) {
+        var resourceType = resourceNode.get("resourceType");
+        if (resourceType != null
+            && "PATIENT".equalsIgnoreCase(resourceType.asText())
+            && theNode.has("fullUrl")) {
+          fullUrl = theNode.get("fullUrl").asText();
+        }
+      }
+    }
+    return fullUrl;
+  }
+
+  public static String updateFullUrl(
+      final String fullUrl, final String oldPatientId, final String newPatientId) {
+    if (!StringUtils.isBlank(fullUrl) && fullUrl.endsWith(oldPatientId)) {
+      return fullUrl.substring(0, fullUrl.length() - oldPatientId.length()) + newPatientId;
+    }
+    return fullUrl;
+  }
+
+  public static String replacePatientRefs(String json, String oldPatientId, String newPatientId) {
+    final String pattern = "(?i)(\"reference\"\\s*:\\s*\"Patient/" + oldPatientId + "\")";
+    return json.replaceAll(pattern, "\"reference\": \"Patient/" + newPatientId + "\"");
+  }
+
+  public static String replacePatientRefs(String json, String newPatientId) {
+    final String pattern = "(?i)(\"reference\"\\s*:\\s*\"Patient\\/[A-Za-z0-9\\-\\.]{1,64}\")";
+    return json.replaceAll(pattern, "\"reference\": \"Patient/" + newPatientId + "\"");
+  }
+
+  public static String replaceFullUrlRefs(String json, String oldFullUrl, String newPatientId) {
+    final String pattern = "(\"reference\"\\s*:\\s*\"" + Pattern.quote(oldFullUrl) + "\")";
+    return json.replaceAll(pattern, "\"reference\": \"Patient/" + newPatientId + "\"");
```

Comment:
To my knowledge Full Urls are not a references to patient, or am I misunderstanding this piece of code.

Reply:
ooh there could be a references to full Url, Wasn't aware of that

Reply:
yeah exactly - a reference could contain a fullUrl, though I never saw any JSON doing that..but still wanted to handle that case before we whack all the old fullUrls like we did with the Patient IDs..